File: RedditPictureReader/app/src/main/java/se/slotte/ludwig/redditpicturereader/picture_feed/data/service/ServiceFactory.java
package se.slotte.ludwig.redditpicturereader.picture_feed.data.service;


import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class ServiceFactory {
    public static String BASE_URL = "https://www.reddit.com";

    //Add Interceptor to every response that we make to the server
    public static OkHttpClient defaultHttpClient = new OkHttpClient.Builder()
            .addInterceptor(
                    new Interceptor() {
                        @Override
                        public Response intercept(Chain chain) throws IOException {
                            Request request = chain.request().newBuilder()
                                    .addHeader("Content-type", "application/json").build();
                            return chain.proceed(request);
                        }
                    }).build();


    public static Retrofit createRetrofitService(final String endPoint) {

        final Retrofit retrofit = new Retrofit.Builder()
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(endPoint).client(defaultHttpClient)
                .build();

        return retrofit;
    }
}

File: RedditPictureReader/app/src/main/java/se/slotte/ludwig/redditpicturereader/picture_feed/data/service/ServiceFactory.kt
package se.slotte.ludwig.redditpicturereader.picture_feed.data.service


import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

const val BASE_URL = "https://www.reddit.com"

object ServiceFactory {

    //Add Interceptor to every response that we make to the server
    private val defaultHttpClient = OkHttpClient.Builder()
            .addInterceptor { chain ->
                val request = chain.request().newBuilder()
                        .addHeader("Content-type", "application/json").build()
                chain.proceed(request)
            }.build()


    fun createRetrofitService(endPoint: String): Retrofit {

        return Retrofit.Builder()
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .baseUrl(endPoint).client(defaultHttpClient)
                .build()
    }
}